src/index.ts

```diff
@@ -19,37 +19,55 @@ const run: Edkt['run'] = async (config) => {
     if (!hasConsent) return;
   }
 
-  // This is a no-op if undefined, equals current value or lesser than 0
+  // This is a no-op if undefined or lesser than 0
   viewStore.setStorageSize(featureStorageSize);
   viewStore.insert(pageFeatures, pageMetadata);
 
-  const matchedAudiences = audienceDefinitions
+  const currentTS = timeStampInSecs();
+
+  const newMatchedAudiences = audienceDefinitions
+    // drop prev matched audiences with same version
     .filter((audience) => {
-      return !matchedAudienceStore.matchedAudienceIds.includes(audience.id);
+      return !matchedAudienceStore.matchedAudiences.some(
+        ({ id, version }) => audience.id === id && version === audience.version
+      );
     })
+    // translate audience definitions into engine queries
     .map((audience) => {
       return {
         ...audience,
         conditions: engine.translate(audience),
       };
     })
-    .map((audience) => {
-      const currentTS = timeStampInSecs();
+    // check if query matches any pageViews
+    .map((query) => {
       const pageViewsWithinLookBack = viewStore.pageViews.filter((pageView) => {
-        return audience.lookBack === 0
+        return query.lookBack === 0
           ? true
-          : pageView.ts > currentTS - audience.lookBack;
+          : pageView.ts > currentTS - query.lookBack;
       });
       return {
-        id: audience.id,
+        id: query.id,
+        version: query.version,
         matchedAt: currentTS,
-        expiresAt: currentTS + audience.ttl,
+        expiresAt: currentTS + query.ttl,
         matchedOnCurrentPageView: true,
-        matched: engine.check(audience.conditions, pageViewsWithinLookBack),
+        matched: engine.check(query.conditions, pageViewsWithinLookBack),
       };
     })
+    // keep only matched audiences
     .filter((audience) => audience.matched);
 
+  // We also want to keep prev matched audiences with matching versions
+  const prevMatchedAudiences = matchedAudienceStore.matchedAudiences.filter(
```

Comment:
is this the reverse op of what we do [upfront](https://github.com/AirGrid/edgekit/pull/130/files#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80R31) - do you think it would be good to combine them? 

the less logic/looping the less change of a bug?

Reply:
Like a `partition`?
Let me have a take on this! Just a minute.